### ISSUE TITLE
Capturing paste event is problematic on mac

### ISSUE THREAD

**Issue by fenec** [mapbox/mapbox-gl-geocoder#34]
Would be good to catch a `paste` event on geocoder input

**Comment by tmcw** [comment]
I'm not sure why this would be necessary, can you specify an example usecase for listening to `paste`? There are _many_ events that any input emits, and individually passing each event without a clear usecase will make this control's code very convoluted.

**Comment by fenec** [comment]
@tmcw sure. the geocoder simply won't respond in any way if a user pastes address from clipboard. 

btw, examples seem to be broken: https://www.mapbox.com/mapbox-gl-js/example/mapbox-gl-geocoder/ 
I cannot see geocoder input

**Comment by andrewharvey** [comment]
@fenec The examples are fixed now.

On Chrome, pasting with Ctrl+V or middle mouse (linux) will ensure the suggestion list is opened, however pasting with right click "Paste", doesn't open the suggestion list. Not sure if the `paste` event helps though, since I couldn't find the actual text contents in the paste event and the input is only filled in after the paste event is called.

**Comment by andrewharvey** [comment]
fix proposed at https://github.com/tristen/suggestions/pull/17

**Comment by edkim** [comment]
@andrewharvey Now that [tristen/suggestions#17](https://github.com/tristen/suggestions/pull/17) has been merged can we incorporate into the geocoder input?

On Mac, pasting into the geocoder search using CMD+V doesn't trigger the suggestions list. I have to press delete or another key to see a list of matching locations (CTRL works, but arrow keys don't).

**Comment by andrewharvey** [comment]
@edkim not yet. We need to do a release of suggestions (https://github.com/tristen/suggestions/issues/19) with the latest changes so the dependency can be updated here.

**Comment by andrewharvey** [comment]
@edkim I did a patch release of `suggestions` which included the bug fix for this. npm will pick up the latest patch release even without bumping the version here in package.json (not sure if we should bump it regardless). So simply redoing `npm install` followed by a rebuild should bring in those fixes.

Would be good to see a new release of `mapbox-gl-geocoder` to bring in the `suggestions` fixes and the `language` param.

**Comment by ogdenstudios** [comment]
Hey everyone, this is still a bug as of v4.2.0. Using mapbox-gl-geocoder v4.2.0 with mapbox-gl-js v1.11.1. I had a handful of user bug reports that pasting addresses was not working. Verified it on OSX with Firefox 79. 

Had to fix it quick so I added a hack to fire off a phony `keydown` event: 

```
var addressSearch = document.querySelector('.mapboxgl-ctrl-geocoder--input');
addressSearch.addEventListener('paste', function(event) {
  var artificialEvent = new Event('keydown');
  event.target.dispatchEvent(artificialEvent);
})
```

This basically fixed it for my usecases, but I think you've still got an issue here.

**Comment by vandres** [comment]
Can confirm, that our Mac users are also confused. CTRL+V on WIndows works though, so for me this looks like a Bug

**Comment by andrewharvey** [comment]
From the reports this is specific to Mac,but seems like it's cross browser.

I don't have a mac so can't test this but the code which captures a paste event is

https://github.com/mapbox/mapbox-gl-geocoder/blob/a5591a3cab9e3d133aeff99d4f31a659f9d9d614/lib/index.js#L295-L300

Perhaps for privacy or other reasons something is stopping access to `e.clipboardData || window.clipboardData`, so text is pasted but not available to access through the keyboard, and `keyDown` isn't triggered. This is just my speculation.

One potential fix is on the `geocode` event by sources the input value directly instead of through the clipboard. This is essentially what @ogdenstudios's workaround does.

Happy to look at a PR that addresses this, but of course this will need help from people with a mac to test it on.

**Comment by ogdenstudios** [comment]
Thanks, Andrew! 

Let me see if I can adapt my workaround a little smarter over the weekend. I can test it out on my mac and I'll submit a PR. Feel free to assign this issue to me if that's how you do things around here.

**Comment by ogdenstudios** [comment]
Hey @andrewharvey - I've been trying to get to this but running into an unrelated bug. Maybe I've got my dev setup wrong. Here's what I'm at: 

1. Clone the repo
2. `npm install`
3. `npm start`
4. Hit http://localhost:9966
5. Set my API token in localStorage 
6. When I start typing in the Geocoder, I receive the following error: 

```
Error is TypeError: externalGeocoderRes.then is not a function
```

Happens across Firefox, Chrome, on my Macbook, on my gaming PC, and I've tried with my personal token + my employer's token, still get the same problem. 

Did I miss a step here?

**Comment by andrewharvey** [comment]
@ogdenstudios looks like a bug in a recent commit. You can try applying my fix in #390 while developing.

**Comment by ogdenstudios** [comment]
Ayy, looks like this original issue *is* resolved in v4.7.0, but the [Mapbox docs](https://docs.mapbox.com/help/tutorials/local-search-geocoding-api/#add-the-geocoder) are only referencing the CDN for v4.2.0.

I tried to hit `<script src='https://api.mapbox.com/mapbox-gl-js/plugins/mapbox-gl-geocoder/v4.7.0/mapbox-gl-geocoder.min.js'></script>` but got no response there. So I copied the current master branch into my project and the `paste` event worked without any workaround. 

I'm a little new to the plugin ecosystem here. How closely tied are you to the Mapbox docs team? Should I file a report with them or are you able to push a new release up there?

**Comment by andrewharvey** [comment]
@ogdenstudios https://github.com/mapbox/mapbox-gl-js-docs/pull/391 is updating the examples and they noticed the latest release hasn't been uploaded to the CDN, so it should be resolved at some point, in the meantime to use 4.7.0 you'll need to host yourself, but versions older than 4.7.0, but newer than 4.2.0 should be available.

I don't think the tutorial is covered by the https://github.com/mapbox/mapbox-gl-js-docs repo, but you could probably still open an issue there so the docs team is aware of it.

The fix for the paste event was done in https://github.com/mapbox/mapbox-gl-geocoder/blob/master/CHANGELOG.md#v450

If you're not seeing this issue on master, I'll close this ticket, but if it's still any issue for anyone on the latest version feel free to post back here or open a new ticket.
